# Introducao e Exmplos

In [15]:
test = [5, 1, 2, 3, 4]

print(max(test))
print(min(test))
print(sum(test))

5
1
15


In [3]:
print(isinstance("Python", (int,str)))
print(isinstance(10.5,int))

True
False


In [16]:
print(list(filter(lambda x: x > 2, test)))

print(list(map(lambda x: x * 2, test)))

[5, 3, 4]
[10, 2, 4, 6, 8]


In [22]:
print(list(reversed(test)))

print(list(sorted(test)))

print(list(sorted(test, reverse = 1)))

[4, 3, 2, 1, 5]
[1, 2, 3, 4, 5]
[5, 4, 3, 2, 1]


## Funcao recursiva

São aquelas funções que chamam a si mesmas para resolver problemas
repetitivos até atingir uma condição de parada

In [23]:
def factorial(n):
    if n == 0:
        return 1
    return n * factorial(n - 1)

print(factorial(5))

120


### GPT explanation of the execution of the function above

First call: factorial(5) is called. Since n is not 0, it moves to return 5 * factorial(4).

But before it can multiply, it needs to know what factorial(4) is, so it calls factorial(4).

Second call: factorial(4) is now called. Again, n isn't 0, so it returns 4 * factorial(3).

Again, it needs to know what factorial(3) is, so it calls factorial(3).

Third call: factorial(3) is called. It returns 3 * factorial(2), so it calls factorial(2).

Fourth call: factorial(2) is called. It returns 2 * factorial(1), so it calls factorial(1).

Fifth call: factorial(1) is called. It returns 1 * factorial(0), so it calls factorial(0).

Sixth call: factorial(0) is called. Now, the base case is hit: it immediately returns 1, because if n == 0: return 1.

---

Now that factorial(0) returns 1, the chain of unfinished multiplications starts resolving:

factorial(1) now becomes 1 * 1 = 1.

factorial(2) now becomes 2 * 1 = 2.

factorial(3) now becomes 3 * 2 = 6.

factorial(4) now becomes 4 * 6 = 24.

factorial(5) now becomes 5 * 24 = 120.

Finally, print(factorial(5)) prints 120.

---

To answer your question:
Yes, once n reaches 0 and returns 1, it "unwinds" through all the previous incomplete multiplications, one by one, finishing the chain backwards.

## Lambda

In [24]:
soma = lambda a, b: a + b

print(soma(3,5))

8


## Funcoes com closures

Closure é uma função interna que lembra do ambiente onde foi criada, mesmo
após a execução da função externa ter terminado. Isso permite que ela mantenha
o estado de variáveis sem a necessidade de usá-las globalmente.

In [25]:
def mulitplicador(n):
    def multiplicado(x):
        return x * n
    return multiplicado

triplo = mulitplicador(3)
valor = triplo(5)
print(valor)

15


## Ex 01

In [27]:
def age():
    b_year = int(input("Insert the birth year: "))
    c_year = int(input("Insert the current year: "))
    age = c_year - b_year
    print(f"They are {age} years old.")
    
age()

They are 27 years old.


2nd Idea

In [ ]:
def age(year1, year2):
    return abs(year1 - year2)

input_01, input_02 = (input("Insert the current year and the year of birth: ").split())

print(age(int(input_01),int(input_02)))

27


## Ex 02

In [29]:
def letter_count(word):
    return len(word)

w_input = input("Insert the word you want to know the amount of letters in: ")
length = letter_count(w_input)

print(f"Your word has {length} letters")

Your word has 6 letters


2nd idea, without using len()

In [ ]:
def word_length(word):
    n = 0
    b = True

    while b:
        try:
            b = bool(word[n])
        except:
            b = False
            break
        n += 1
    return n

input_w = word_length(input("Insert your word: "))

print("Your word has", input_w,"words")

Your word has 7 words


## Ex 03

In [51]:
def greetings(time):
    if time < 12:
        return "Bom dia!"
    elif 12 <= time < 18:
        return "Boa Tarde!"
    else:
        return "Boa Noite!"
    
time = int(input("Insert what time it is (24h format): "))

print(greetings(time))

Boa Noite!


## Ex 04

In [65]:
telefones = ["11987654321", "21912345678", "31987654321", "11911223344"] 

def convert_str2int(num_list):
    for n in range(len(num_list)):
        num_list[n] = int(num_list[n])
    return num_list


def isallint(num_list):
    check = False
    for n in range(len(num_list)):
        if isinstance(num_list[n], int):
            check = True
        else:
            check = False
            break
    if check:
        print("All numbers were converted properly.")
    else:
        print("Number conversions failed.")
        

telefones = convert_str2int(telefones)
print(telefones)

isallint(telefones)

[11987654321, 21912345678, 31987654321, 11911223344]
All numbers were converted properly.


### Provided solution

In [ ]:
def converter(list):
    return [ int(tel) for tel in list]

def verify_nums(list):
    for num in list:
        if not isinstance(num, int):
            return "Error in conversion"
        return "All numbers were converted correctly."

telefones = ["11987654321", "21912345678", "31987654320", "11911223344"] 

converted_numbers = converter(telefones)

print(verify_nums(converted_numbers))    

All numbers were converted correctly.


## Ex 05

In [76]:
def total_sales(sales):
    sales = [float(sale) for sale in sales]
    return sum(sales)

sales = input("Insert the sales of each day (separated by spaces ' '): ").split()

print("The total sales were:", total_sales(sales))

The total sales were: 650.0


### Provided Solution

In [79]:
values = input("Insert the daily sales: ").split()
total = sum(map(float, values))
print(f"The total sales is: {total:.2f}")

The total sales is: 650.00


## Ex 06

In [84]:
values = input("Insert the numbers separated by space: ").split()
values = [int(n) for n in values]
even = list(filter(lambda x: x % 2 == 0, values))
print(f"Even numbers: {even}")


Even numbers: [2, 4, 6]


### Provided Solution

In [86]:
numbers = input("Insert the numbers separated by space: ").split()
even = filter(lambda x: int(x) % 2 == 0, numbers)
print("Even numbers:", ", ".join(even))

Even numbers: 2, 4, 6


## Ex 07

In [87]:
products = input("Insert the product names separated by commas ',': ").split(", ")
prices = input("Insert the prices separated by commas ',': ").split(", ") 

pp_list = list(zip(products,prices))

for pp in pp_list:
    print(": ".join(pp))

banana: 1.2
apples: 2.3
mangos: 8.5


### Provided solution

In [88]:
products = input("Insert the product names separated by commas ',': ").split(",")
prices = input("Insert the prices separated by commas ',': ").split(",") 

for products, prices in zip(products, prices):
    print(f"{products.strip()}: {prices.strip()}")

Banana: 1.2
Apples: 2.3
Mangos: 8.5


### Combination Idea

In [89]:
products = input("Insert the product names separated by commas ',': ").split(", ")
prices = input("Insert the prices separated by commas ',': ").split(", ") 

for pp in list(zip(products,prices)):
    print(": ".join(pp))

Banana: 1.2
Apples: 2.3
Mango: 8.5


## Ex 08

In [90]:
def calculator():
    num01 = float(input("Insert the first number: "))
    num02 = float(input("Insert the second number: "))
    mathOp = input("Choose the operation type (| + | - | * | / |): ").strip()
    match mathOp:
        case mathOp if mathOp == "+":
            return num01 + num02
        case mathOp if mathOp == "-":
            return num01 - num02
        case mathOp if mathOp == "*":
            return num01 * num02
        case mathOp if mathOp == "/":
            return num01 / num02
        
print("The result is: ", calculator())

The result is:  15.0


### Exercise asked to create the calculator using lambda

In [91]:
add = lambda x, y: x + y
subtract = lambda x, y: x - y
multiply = lambda x, y: x * y
divide = lambda x, y: x / y


y = float(input("Insert the first number: "))
x = float(input("Insert the second number: "))
mathOp = input("Choose the operation type (| + | - | * | / |): ").strip()

match mathOp:
    case mathOp if mathOp == "+":
        print("The result is: ", add(x,y))
    case mathOp if mathOp == "-":
        print("The result is: ", subtract(x,y))
    case mathOp if mathOp == "*":
        print("The result is: ", multiply(x,y))
    case mathOp if mathOp == "/":
        print("The result is: ", divide(x,y))
    case _:
        print("Invalid input.")

The result is:  15.0


### Provided more eficient solution

In [93]:
add = lambda x, y: x + y
subtract = lambda x, y: x - y
multiply = lambda x, y: x * y
divide = lambda x, y: x / y

Operations = {
    '+': add,
    '-': subtract,
    '*': multiply,
    '/': divide
}

y = float(input("Insert the first number: "))
x = float(input("Insert the second number: "))
mathOp = input("Choose the operation type (| + | - | * | / |): ").strip()

if mathOp in Operations:
    result = Operations[mathOp](x, y)
    print(f"The result is: {result}")
else:
    print("Invalid input.")

The result is: 50.0


## Ex 09

In [ ]:
def discount(value):
    def discounted(price):
        return price - (price * value)
    return discounted

disc_amount = float(input("Insert the percentage value of the discount (without the % symbol): ")) / 100
discounter = discount(disc_amount)
final_price = discounter(float(input("Insert the value of the product to be discounted: ")))

print(f"The final discounted price is: {final_price:.2f}")

The final discounted price is: 180.00


Matched the solution given on first try :b

## Ex 10

In [ ]:
def addfall(n):
    if n == 1:
        return 1
    return n + addfall(n -1)

n = int(input("Insert a whole number: "))
print(f"The sum from 1 to {n} is: {addfall(n)}")

The sum from 1 to 5 is: 15


Likewise matched the solution given first try (they had given a very similar example at the beginning)